Multiple Dashboards

stefwalter opened the pull request:
This does the navigation refactoring and redesign noted in #1726. 

 * Move to a navigation more compatible with Patternfly
 * Multiple dashboards are now possible
 * The navbar and sidebar are implemented in pkg/shell/index.html and all the routing, loading of components, listing of machines for menus and related logic happens in pkg/shell/index.js
 * All server pages are loaded as components, even the ones in shell/shell.html
 * #hash URL formats change.

See below for behavioral change overview.

stefwalter commented:
Follow ups:

https://github.com/cockpit-project/cockpit/issues/1865
https://github.com/cockpit-project/cockpit/issues/1859

For now the kubernetes dashboard will be hard coded in the top bar, but I would like to resolve #1859 next week and make it appear only when the kubernetes is installed.

stefwalter commented:
Follow up: https://github.com/cockpit-project/cockpit/issues/1867

stefwalter commented:
This is ready for review. Integration tests still need to be updated.

stefwalter commented:
Design work took place here: https://github.com/cockpit-project/cockpit/issues/1726

stefwalter commented:
Design was done together with @andreasn and noted in #1726. A few tweaks were made during implementation, once it turned out that some of the ideas didn't work.

Screenshot tour:

When you first use Cockpit on a machine, you are taken to the system summary page. You can see the host name in the navbar along with the server's avatar.

![screenshot from 2015-02-27 17 35 55](https://cloud.githubusercontent.com/assets/795070/6416532/18906622-bea7-11e4-95c7-d0f382f86604.png)

You can click on 'Dashboard' to go to the machines dashboard, where you can add a machine.

![screenshot from 2015-02-27 17 38 19](https://cloud.githubusercontent.com/assets/795070/6416587/6df69334-bea7-11e4-86c9-d658a763ae67.png)

Note that the host name now changes to a 'Machines' drop down. Once you have more than one machine in Cockpit, you can always use that menu to switch between machines. The colors and avatars that you see are those chosen in the dashboard (as you could before).

![screenshot from 2015-02-27 17 39 02](https://cloud.githubusercontent.com/assets/795070/6416601/8776eb88-bea7-11e4-8507-c6b1d0082239.png)

Once you have more than one machine, you'll start to see the machine's color on the navbar to clearly give context when you select various machines. This machine has a blue color chosen.

![screenshot from 2015-02-27 17 30 51](https://cloud.githubusercontent.com/assets/795070/6416425/68a72408-bea6-11e4-8784-37e5c62eca45.png)

And here case we've selected another machine, which is tagged green. (ignore the "Not Found" here, the other machine didn't have the new package stuff installed).

![screenshot from 2015-02-27 17 40 10](https://cloud.githubusercontent.com/assets/795070/6416621/b04035ba-bea7-11e4-91b6-6c448e209954.png)

This Color + Host name + avatar change is extremely important. If you change between machines in the drop down, you remain in the same section on the new host (eg: storage -> storage) and we need to have clear indications the machine you are operating has changed. Thus the color + avatar change in the navbar.

Obviously, additional dashboards are available in the main navbar. I've hard coded the 'Cluster' link there for now, but soon this will be driven by addition and removal of the cockpit-kubernetes package. See https://github.com/cockpit-project/cockpit/issues/1859 for the work on that.

In the interim i've hard coded our kubernetes tool as a dashboard. This will soon be replaced by the real kubernetes dashboard in https://github.com/cockpit-project/cockpit/issues/1687.

![screenshot from 2015-02-27 17 45 24](https://cloud.githubusercontent.com/assets/795070/6416710/6afef60c-bea8-11e4-8c84-ce1ca5f3bb74.png)

stefwalter commented:
Follow up: https://github.com/cockpit-project/cockpit/issues/1868

stefwalter commented:
This requires reinstallation of cockpit (ie: make install) and restarting cockpit-ws. If you're using multiple machines, all the machines need to be updated.

stefwalter commented:
More follow ups:

 * Use the new patternfly: https://github.com/cockpit-project/cockpit/issues/1870
 * Have a spinner on far right of navbar while loading a component page: https://github.com/cockpit-project/cockpit/issues/1871

stefwalter commented:
* Fixed 'Account Settings' menu item.
 * Pass options to frames
 * check-login passes

stefwalter commented:
All tests pass except for some check-storage races.

mvollmer commented:
#1876 should help with the storage races.

stefwalter commented:
> #1876 should help with the storage races.

Yup. Thanks!

mvollmer commented:
This seems to be genuine.  I'll investigate.

```
======================================================================
ERROR: testBasic (__main__.TestRoles)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./check-roles", line 104, in testBasic
    b.wait_val("#account-real-name", "Administratorrrr")
  File "/home/mvo/work/cockpit/test/testlib.py", line 215, in wait_val
    return self.wait_js_func('ph_has_val', selector, val)
  File "/home/mvo/work/cockpit/test/testlib.py", line 203, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))), timeout=self.phantom_wait_timeout)
  File "/home/mvo/work/cockpit/test/testlib.py", line 534, in wait
    return self.run({'cmd': 'wait', 'cond': cond, 'timeout': timeout*1000})
  File "/home/mvo/work/cockpit/test/testlib.py", line 500, in run
    raise Error("timeout")
Error: timeout
```

mvollmer commented:
`check-roles` tried to change the real name of "admin" but ended up changing the real name of "user" because the parameters of the page changed asynchronously from "account?id=admin" back to "account?id=user" during the test run, probably because of the relogin.

The sequence is something like this:

```
b.go("account?id=user")
b.enter_page("account")
b.relogin("account")
b.go("account?id=admin")
b.enter_page("account")            // was missing, but isn't enough
b.wait_text("#account-user-name", "admin")
// now we are on the "id=admin" page
// something happens in the background and we go back to "id=user"
b.set_val("#account-real-name", "Administratorrrr")
```

I could fix this like this:

```
diff --git a/test/check-roles b/test/check-roles
index 9f263c3..5da2710 100755
--- a/test/check-roles
+++ b/test/check-roles
@@ -85,9 +85,11 @@ class TestRoles(MachineCase):
         b.wait_text("#account-roles", "Server Administrator")
 
         b.relogin("account", "user")
+        b.wait_text("#account-user-name", "user")
 
         # Try to change admins name again
         b.go("account?id=admin")
+        b.enter_page("account")
         b.wait_text("#account-user-name", "admin")
         b.wait_present("#account-real-name:not([disabled])")
         b.wait_present("#account-locked:not([disabled])")
```

@stefwalter, does this make sense?  (I'll go read the code now...)

mvollmer commented on pkg/shell/cockpit-docker.js:
```diff
@@ -1699,8 +1697,7 @@ PageContainerDetails.prototype = {
 
-        this.address = shell.get_page_machine();
-        this.client = shell.docker(this.address);
+        this.client = shell.docker(null);
```
Same.

mvollmer commented on pkg/shell/cockpit-docker.js:
```diff
@@ -2286,7 +2282,7 @@ function DockerClient(machine) {
 
         /* TODO: This code needs to be migrated away from dbus-json1 */
-        dbus_client = shell.dbus(machine);
+        dbus_client = shell.dbus();
```
In other places, this is `shell.dbus(null)`.

mvollmer commented on pkg/shell/cockpit-docker.js:
```diff
@@ -1409,8 +1408,7 @@ PageSearchImage.prototype = {
     enter: function() {
-        this.address = shell.get_page_machine();
-        this.client = shell.docker(this.address);
+        this.client = shell.docker(null);
```
`shell.docker` doesn't take any arguments anymore.

stefwalter commented:
Fixed 2 shell.docker(null) issues.

mvollmer commented:
Reloading a page appends `/dashboard/dashboard/list` to the URL, which leads to a "Not Found" error. Reloading a couple of times adds more `dashboard` strings.
![screenshot from 2015-03-03 11 59 47](https://cloud.githubusercontent.com/assets/3228183/6460083/f45c07a4-c19c-11e4-862c-2ea9235abd7e.png)

stefwalter commented:
Fix for adding too many ```dashboard/list```

```
-                cockpit.location.go("dashboard/list");
+                cockpit.location.go("/dashboard/list");
```

mvollmer commented:
The components for a remote machine are actually showing stuff for localhost.

Note that this should show "Dandelion" as the hostname but shows "Sunflower", which is localhost.

![screenshot from 2015-03-03 12 06 10](https://cloud.githubusercontent.com/assets/3228183/6460202/d3fd741a-c19d-11e4-8653-119755459a5a.png)

stefwalter commented:
Fixed issue where reloading would go to dashboard. The ```navigate()``` function was being called before local machines were loaded.

Fixed issue where shell.html was connecting to wrong machines for ```shell.dbus(null)``` calls.

mvollmer commented:
Reloading loses subnavigation, but I think that's OK for now.  (E.g., navigate to a network interface, reload, you are back at the network overview.)

mvollmer commented:
> Fixed issue where reloading would go to dashboard. The navigate() function was being called before 
> local machines were loaded.

Works.

> Fixed issue where shell.html was connecting to wrong machines for shell.dbus(null) calls.

Works.

stefwalter commented:
> Reloading loses subnavigation, but I think that's OK for now. (E.g., navigate to a network interface, reload, you are back at the network overview.)

This is only the case on shell legacy pages. And yes, that's an unfortunate limitation, but we should be migrating away from that.

stefwalter commented:
Follow ups tracked here: https://trello.com/c/59Ajm8lz/53-multiple-dashboards
